### PR TITLE
[REEF-1873] Driver side interfaces: service set up and failures

### PR DESCRIPTION
**Pull request by interesaaat**
This addressed the issue by
  * adding Driver interfaces for the service, subscriptions and task sets.
  * adding Failures interfaces.

JIRA:
  [REEF-1873](https://issues.apache.org/jira/browse/REEF-1873)

Pull request:
  This closes #

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Network/Elastic/Driver/IElasticTaskSetService.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+using Org.Apache.REEF.Network.Elastic.Failures;
+using Org.Apache.REEF.Tang.Interface;
+
+namespace Org.Apache.REEF.Network.Elastic.Driver
+{
+    /// <summary>
+    /// Used to create Subscriptions for fault tolerant Task Sets.
+    /// Also manages configurations for Group Communication operators/services.
+    /// </summary>
+    public interface IElasticTaskSetService : IFailureResponse
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Elastic/Driver/IElasticTaskSetService.cs]: Please mark all new interfaces with `@Unstable` and set the appropriate fields inside the annotation.

File: lang/cs/Org.Apache.REEF.Network/Elastic/Driver/ITaskSetManager.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+// under the License.
+
+using System.Collections.Generic;
+using Org.Apache.REEF.Driver.Context;
+using Org.Apache.REEF.Tang.Interface;
+using Org.Apache.REEF.Driver.Evaluator;
+using Org.Apache.REEF.Driver.Task;
+using Org.Apache.REEF.Network.Elastic.Failures;
+
+namespace Org.Apache.REEF.Network.Elastic.Driver.Impl
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Elastic/Driver/ITaskSetManager.cs]: Why is an interface in the `Impl` name space? Also, the file isn't in the `Impl` folder.

File: lang/cs/Org.Apache.REEF.Network/Elastic/Failures/IFailureEvent.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+using System;
+
+namespace Org.Apache.REEF.Network.Elastic.Failures
+{
+    /// <summary>
+    /// Interface wrapping an event rised by a transition to a new failure
+    /// state. The event speicifies which action have to be executed in response
+    /// to the change in the failure state.
+    /// </summary>
+    public interface IFailureEvent : IDisposable
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Elastic/Failures/IFailureEvent.cs]: Why is a Failure `IDisposable`? If it has to be, please document the `.Dispose` method in here.

File: lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj [hunk 4]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj]: if possible, please remove the redundant change here.

File: lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj [hunk 5]
```diff
@@ -216,7 +223,8 @@ under the License.
       <Name>Org.Apache.REEF.Wake</Name>
     </ProjectReference>
   </ItemGroup>
+  <ItemGroup />
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj]: Please remove.

File: lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj [hunk 6]
```diff
@@ -216,7 +223,8 @@ under the License.
...
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
   <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
   <Import Project="$(PackagesDir)\StyleCop.MSBuild.$(StyleCopVersion)\build\StyleCop.MSBuild.Targets" Condition="Exists('$(PackagesDir)\StyleCop.MSBuild.$(StyleCopVersion)\build\StyleCop.MSBuild.Targets')" />
-</Project>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj]: If possible, please remove this change.

File: lang/cs/Org.Apache.REEF.Network/Elastic/Driver/ITaskSetManager.cs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        /// </summary>
+        /// <param name="evaluator">The failed evaluator</param>
+        void OnEvaluatorFailure(IFailedEvaluator evaluator);
+
+        /// <summary>
+        /// Contains the logic to trigger when the execution fails.
+        /// </summary>
+        void OnFail();
+
+        void Dispose();
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Elastic/Driver/ITaskSetManager.cs]: please document.

File: lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj (hunk 4 above)
Comment (interesaaat) [reply lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj]: No idea how to fix this.

Comment (motus) [reply lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj]: Most likely the difference is the Unicode 0xFEFF marker inserted (or removed?) by the VisualStudio. You can open the file in binary mode e.g. in `vim -b` and fix that if you want.

File: lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj [hunk 8]
```diff
@@ -219,4 +226,4 @@ under the License.
...
+</Project>
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj]: Same here, by the way.

File: lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj (hunk 4 above)
Comment (interesaaat) [reply lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj]: How does it look now?

Comment (motus) [reply lang/cs/Org.Apache.REEF.Network/Org.Apache.REEF.Network.csproj]: Looks good now!

**Comment by motus** [comment]
I have a question, though - why do we have a PR with interfaces only? Do we plan to submit a separate PR that provides the implementation? I find it a bit hard to review an abstract API without really understanding how it will work and in what context.

**Comment by interesaaat** [comment]
@motus, how do you prefer me to PR my code? I was planning to first PR all driver-related interfaces, then upload the code. Subsequently the task-related interface, then the code. I have no experience on this, so please advise on what would be the preferred way to go.

**Pull request closed without merging**